### PR TITLE
ssl: Refactor pre TLS-1.3 to separate client/server state machines 

### PR DESCRIPTION
**Pull request by IngelaAndin**
Same refactor as previously done for TLS-1.3

### CODE REVIEW

**Review by u3s — COMMENTED** [review]
please re-request me as reviewer, when this PR is again ready for review.
this will make my life easier, thank you :-)

**Review by u3s — APPROVED** [review]

File: lib/ssl/src/tls_dtls_client_connection.erl [hunk 1]
```diff
@@ -0,0 +1,869 @@
...
+-include("ssl_connection.hrl").
+-include("ssl_handshake.hrl").
+-include("ssl_alert.hrl").
+-include("ssl_record.hrl").
+-include("ssl_cipher.hrl").
+-include("ssl_internal.hrl").
+-include("ssl_srp.hrl").
+
+
+%% General state handlingfor TLS-1.0 to TLS-1.2
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_client_connection.erl:45]: fixed

File: lib/ssl/src/tls_dtls_gen_connection.erl [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+-export([initial_state/8,
+         negotiated_hashsign/4,
+         finalize_handshake/3,
+         make_premaster_secret/2,
+         handle_peer_cert/6,
+         calculate_master_secret/2,
+         calculate_master_secret/5
+        ]).
+
+%% General state handlingfor TLS-1.0 to TLS-1.2
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_gen_connection.erl:61]: will fix

File: lib/ssl/src/tls_gen_connection.erl [hunk 3]
```diff
@@ -282,6 +258,28 @@ setopts(Transport, Socket, Other) ->
...
+    try
+        handle_info(Event, StateName, State)
+    catch
+        _:Reason:ST ->
+            ?SSL_LOG(info, internal_error, [{error, Reason}, {stacktrace, ST}]),
+	    ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?INTERNAL_ERROR,
+						       malformed_data),
+					    StateName, State)
+    end;
+
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_gen_connection.erl:271]: Will remove

File: lib/ssl/src/tls_server_connection_1_3.erl [hunk 4]
```diff
@@ -118,12 +118,6 @@ init([?SERVER_ROLE, Sender, Host, Port, Socket, Options,  User, CbInfo]) ->
             gen_statem:enter_loop(?MODULE, [], config_error, EState)
     end.
 
-terminate({shutdown, {sender_died, Reason}}, _StateName,
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_server_connection_1_3.erl:121]: This is dead code left over from before we used significant child.

File: lib/ssl/src/tls_client_connection.erl [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+
+%%--------------------------------------------------------------------
+-spec wait_stapling(gen_statem:event_type(), term(), #state{}) ->
+          gen_statem:state_function_result().
+%%--------------------------------------------------------------------
+wait_stapling(info, Event, State) ->
+    tls_gen_connection:gen_info(Event, wait_stapling, State);
+wait_stapling(Type, Event, State) ->
+    try gen_state(wait_stapling, Type, Event, State)
+    catch throw:#alert{} = Alert ->
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_client_connection.erl:361]: Yes, it is. I will fix.

File: lib/ssl/src/tls_dtls_server_connection.erl [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+
+-include("ssl_api.hrl").
+-include("tls_connection.hrl").
+-include("ssl_connection.hrl").
+-include("ssl_handshake.hrl").
+-include("ssl_alert.hrl").
+-include("ssl_record.hrl").
+-include("ssl_cipher.hrl").
+-include("ssl_internal.hrl").
+-include("ssl_srp.hrl").
```
Comment (dgud) [inline lib/ssl/src/tls_dtls_server_connection.erl:42]:
```suggestion
-include_lib("public_key/include/public_key.hrl").

-include("ssl_connection.hrl").
-include("ssl_handshake.hrl").
-include("ssl_alert.hrl").
-include("ssl_record.hrl").
-include("ssl_internal.hrl").
-include("ssl_srp.hrl").
```

File: lib/ssl/src/tls_gen_connection.erl [hunk 7]
```diff
@@ -180,37 +181,11 @@ initialize_tls_sender(#state{static_env = #static_env{
...
-    HelloRequest = ssl_handshake:hello_request(),
-    Frag = tls_handshake:encode_handshake(HelloRequest, Version),
-    Hs0 = ssl_handshake:init_handshake_history(),
-    {BinMsg, ConnectionStates} = 
-	tls_record:encode_handshake(Frag, Version, ConnectionStates0),
-    tls_socket:send(Transport, Socket, BinMsg),
-    State = State0#state{connection_states = 
-			     ConnectionStates,
-			 handshake_env = HsEnv#handshake_env{tls_handshake_history = Hs0}},
-    next_event(hello, no_record, State, Actions).
```
Comment (dgud) [inline lib/ssl/src/tls_gen_connection.erl:208]: This is called from other modules ?

File: lib/ssl/src/tls_server_connection.erl [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+-include_lib("public_key/include/public_key.hrl").
+-include_lib("kernel/include/logger.hrl").
+
+-include("tls_connection.hrl").
+-include("tls_handshake.hrl").
+-include("ssl_alert.hrl").
+-include("tls_record.hrl").
+-include("ssl_cipher.hrl").
+-include("ssl_api.hrl").
+-include("ssl_internal.hrl").
```
Comment (dgud) [inline lib/ssl/src/tls_server_connection.erl:112]:
```suggestion
-include("tls_connection.hrl").
-include("tls_handshake.hrl").
-include("ssl_alert.hrl").
```

File: lib/ssl/src/dtls_client_connection.erl [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+-include_lib("kernel/include/logger.hrl").
+
+-include("dtls_connection.hrl").
+-include("dtls_handshake.hrl").
+-include("ssl_alert.hrl").
+-include("dtls_record.hrl").
+-include("ssl_cipher.hrl").
+-include("ssl_api.hrl").
+-include("ssl_internal.hrl").
+-include("ssl_srp.hrl").
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/dtls_client_connection.erl:129]: I will try

File: lib/ssl/src/tls_dtls_gen_connection.erl [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+%%--------------------------------------------------------------------
+-spec prf(pid(), binary() | 'master_secret', binary(),
+	  [binary() | ssl:prf_random()], non_neg_integer()) ->
+		 {ok, binary()} | {error, reason()} | {'EXIT', term()}.
+%%
+%% Description: use a ssl sessions TLS PRF to generate key material
+%%--------------------------------------------------------------------
+prf(ConnectionPid, Secret, Label, Seed, WantedLength) ->
+    ssl_gen_statem:call(ConnectionPid, {prf, Secret, Label, Seed, WantedLength}).
+
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_gen_connection.erl:103]: backwards compatibility

File: lib/ssl/src/tls_dtls_gen_connection.erl [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+%%--------------------------------------------------------------------
+-spec connection(gen_statem:event_type(), term(), #state{}) ->
+			gen_statem:state_function_result().
+%%--------------------------------------------------------------------
+
+connection({call, From}, renegotiate,
+           #state{static_env = #static_env{protocol_cb = Connection},
+                  handshake_env = HsEnv} = State) ->
+    Connection:renegotiate(State#state{handshake_env =
+                                           HsEnv#handshake_env{renegotiation = {true, From}}}, []);
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_gen_connection.erl:358]: tls_dtls_gen_connetion!

File: lib/ssl/src/tls_gen_connection.erl (hunk 7 above)
Comment (IngelaAndin) [reply lib/ssl/src/tls_gen_connection.erl:208]: moved to tls_dtls_gen_connection

File: lib/ssl/src/dtls_client_connection.erl (hunk 9 above)
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/dtls_client_connection.erl:129]: Removed superfluous includes

File: lib/ssl/src/dtls_server_connection.erl [hunk 12]
```diff
@@ -0,0 +1,574 @@
...
+-include_lib("kernel/include/logger.hrl").
+
+-include("dtls_connection.hrl").
+-include("dtls_handshake.hrl").
+-include("ssl_alert.hrl").
+-include("dtls_record.hrl").
+-include("ssl_cipher.hrl").
+-include("ssl_api.hrl").
+-include("ssl_internal.hrl").
+-include("ssl_srp.hrl").
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/dtls_server_connection.erl:129]: It was correct

File: lib/ssl/src/tls_dtls_client_connection.erl [hunk 13]
```diff
@@ -0,0 +1,869 @@
...
+
+-include("ssl_api.hrl").
+-include("tls_connection.hrl").
+-include("ssl_connection.hrl").
+-include("ssl_handshake.hrl").
+-include("ssl_alert.hrl").
+-include("ssl_record.hrl").
+-include("ssl_cipher.hrl").
+-include("ssl_internal.hrl").
+-include("ssl_srp.hrl").
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_client_connection.erl:42]: fixed

File: lib/ssl/src/tls_dtls_server_connection.erl (hunk 6 above)
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_server_connection.erl:42]: fixed

File: lib/ssl/src/tls_server_connection.erl (hunk 8 above)
Comment (IngelaAndin) [reply lib/ssl/src/tls_server_connection.erl:112]: fixed

File: lib/ssl/src/tls_gen_connection.erl (hunk 7 above)
Comment (IngelaAndin) [reply lib/ssl/src/tls_gen_connection.erl:208]: Seems I was mistaken, will look into it more

File: lib/ssl/src/tls_dtls_gen_connection.erl (hunk 10 above)
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_gen_connection.erl:103]: Oh , I see I misunderstood the question. I think it actually is a little part of the very original design  that kept the calls in the same  module as the handling of the call. But this is no longer a working principle so maybe we can change that.

[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_gen_connection.erl]: The prf has become dead code so I removed it. The renegotiate could be moved, but I think it has the value of linking it to pre TLS-1.3 code so I think I will leave it.

File: lib/ssl/src/tls_dtls_gen_connection.erl (hunk 11 above)
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/tls_dtls_gen_connection.erl]: I realized that it is the DTLS-code that failed to make the split properly. Should be fixed now.

File: lib/ssl/src/tls_gen_connection.erl (hunk 7 above)
Comment (IngelaAndin) [reply lib/ssl/src/tls_gen_connection.erl:208]: See comment above

**Pull request merged**